File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        no = {
            '2':'abc',
            '3':'def',
            '4':'ghi',
            '5':'jkl',
            '6':'mno',
            '7':'pqrs',
            '8':'tuv',
            '9':'wxyz'
        }
        
        n = len(digits)
        result = []
        def backtrack(temp, curr):
            if curr == n :
                result.append(temp)
                return
            for i in digits[curr]:
                for char in no[i]:
                    backtrack(temp + char, curr + 1)
        
        backtrack("", 0)
        return result

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    private String[] arr = new String[8];
    private List<String> result = new ArrayList<>();
    private String num;

    private void backtrack(StringBuilder temp, int curr){
        if(curr == num.length()){
            result.add(temp.toString());
            return;
        }

        int c = Character.getNumericValue(num.charAt(curr));
        for(char ch: arr[c-2].toCharArray()){
            temp.append(ch);
            backtrack(temp, curr+1);
            temp.deleteCharAt(temp.length()-1);
        }
    }

    public List<String> letterCombinations(String digits) {
        if(digits.equals("")) return new ArrayList<>();
        arr[0] = "abc";
        arr[1] = "def";
        arr[2] = "ghi";
        arr[3] = "jkl";
        arr[4] = "mno";
        arr[5] = "pqrs";
        arr[6] = "tuv";
        arr[7] = "wxyz";

        num = digits;
        backtrack(new StringBuilder(), 0);
        return result;
    }
}